### ISSUE TITLE
Eager: Variable item-assignment

### ISSUE THREAD

**Issue by asimshankar** [tensorflow/tensorflow#14132]
(This applies only when [eager execution](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/eager/README.md) has been enabled via `tfe.enable_eager_execution()`)

Currently, the following does not work:

```python
import tensorflow as tf
import tensorflow.contrib.eager as tfe

tfe.enable_eager_execution()
x = tfe.Variable(tf.ones([10]), name=’x’)
x[3] = 2.
```

That last line, assigning a value to an element of the tensor will fail with: `object does not support item assignment`.

While it would be possible to make item assignment for `Tensor` and `Variable` objects work when eager execution is enabled, it is trickier to make the same line of code work with graph construction (since the line `x[3] = 2.` does not return a `tf.Operation` object that can be provided to a `Session.run()` call).

At this early stage of eager execution, we’re taking the conservative approach and disallowing Tensor assignment. This is probably worth revisiting at a future date. In the meantime, the verbose form of item assignment using `tf.scatter_update`:

```python
tf.scatter_update(x, [3], [2.])
```

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 190 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 206 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 221 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 236 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 251 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 266 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 281 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 296 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 311 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 326 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tbenst** [comment]
This is unfortunately a show-stopper for porting code from PyTorch

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
@johnpjust : could you provide a short snippet to reproduce the error you're seeing? From the error message it appears that `scatter_update` is being invoked on a `Tensor` instead of a `Variable`.

**Comment by johnpjust** [comment]
Hi yes -- you are correct in that I am trying to set it on a single element of the variable array, which I thought is like what is happening above.  I've got this on SO as well since i wasn't sure if it was a bug, but when i use scatter_update I'm not finding the right combination to change a single element of, e.g., a 3x3 variable

model.trainable_variables[0]
<tf.Variable 'dense_5/kernel:0' shape=(3, 3) dtype=float64, numpy=
array([[ 0.58309446,  0.47310445,  0.69813648],
       [ 0.2342269 ,  0.01559285, -0.77714921],
       [-0.43583301,  0.45498991,  1.31445808]])>

tf.scatter_update(model.trainable_variables[0], [1,1],3.0) for instance will result in 
<tf.Variable 'UnreadVariable' shape=(3, 3) dtype=float64, numpy=
array([[ 0.58309446,  0.47310445,  0.69813648],
       [ 3.        ,  3.        ,  3.        ],
       [-0.43583301,  0.45498991,  1.31445808]])>

**Comment by veritas9872** [comment]
Hello. I would like to comment on this issue.

In fact, there is a method to get item assignment in both eager mode and graph mode.

If you look at the source code for tf.image.random_flip_up_down (specifically in the function _random_flip), there is a simple way to do what appears to be in-pace assignment (when viewed from far enough away). 

Just make a mask tensor with the indices that you want to replace set to 0 and the locations that you want to use set at 1 (call this tensor "mask").

Then perform whatever operation that you want to do on a tensor that is the same size (or is broadcastable to) your original tensor (call this tensor "other").

Then you can simply create a new tensor by doing the following:

new = original * mask + other * (1 - mask)

This would create a new tensor which appears to have undergone item assignment.

**Comment by huan** [comment]
@veritas9872 It seems like a good workaround! :+1:

**Comment by veritas9872** [comment]
I think that it would be great if Tensorflow could implement this natively at a lower level. However, I don't know how to get the attention of the people who could actually implement this.

**Comment by huan** [comment]
@veritas9872 How about write a PyPI module that does this job by a wrapper?

If there's feature is really a thing, then I believe lots of people will install and use it. :D

**Comment by RasoulNik** [comment]
The following is taken from tensorflow2.0 Alpha documentation:
[link](https://www.tensorflow.org/alpha/tutorials/eager/tf_function)
"So, for example, the best way to accumulate data in a loop is still to use tf.TensorArray:
@tf.function
def f(x):
  ta = tf.TensorArray(tf.float32, size=10)
  for i in tf.range(10):
    x += x
    ta = ta.write(i, x)
  return ta.stack()

f(10.0)"

**Comment by dynamicwebpaige** [comment]
Closing out this issue, as `tf.contrib` is no longer supported in TF 2.0. For eager behavior, please refer to the documentation referenced above by @RasoulNik. Thank you! 🙂

**Comment by ethanyanjiali** [comment]
Tensorflow sucks

**Comment by KhurramPirov** [comment]
In my case,  I had to assign depth_buffer[v,u] = point_depth. Here is the solution: 
depth_buffer = depth_buffer[v:v+1,u:u+1].assign([point_depth])

Tensorflow does not suck on prod))

**Comment by ethanyanjiali** [comment]
@KhurramPirov nah i also use tf for prod but i know clearly that it's just a matter of time for pytorch to dominate at least vision community.

**Comment by gajeshladhar** [comment]
> Tensorflow sucks

Yes , tensorflow realy sucks , people have to solve some un-related problems due to tensorflows mistakes, i dont know how long it will take to launch neat and clean tensorflow version, sometimes thought older 1.x was better that this rubbish tf 2.x

**Comment by darraghdog** [comment]
Was struggling here for a bit... here is what I got, 

```
# pytorch
idx = torch.tensor([1,2])
a = torch.zeros((4, 4))
b = torch.ones((4, 4))
a[idx] += b[idx]
```       

```   
# tensorflow    
idx = tf.constant([1,2])
a = tf.zeros((4, 4))
b = tf.ones((4, 4))
a = tf.tensor_scatter_nd_update(a, tf.expand_dims(idx,1), tf.gather_nd(b, tf.expand_dims(idx,1)))
```

**Comment by K-tang-mkv** [comment]
Firstly change the tensor to numpy array and update it, and then change back to tensorflow tensor.

**Comment by etasnadi** [comment]
> Firstly change the tensor to numpy array and update it, and then change back to tensorflow tensor.

That will break the gradient flow.
